### ISSUE TITLE
Unnessary loads in ReadOnlySpan(T[] array, int start, int length)

### ISSUE THREAD

**Issue by benaadams** [dotnet/coreclr#9633]
The following lines get converted 
```csharp
_pointer = new ByReference<T>(ref Unsafe.Add(ref JitHelpers.GetArrayData(array), start));
_length = length;
```
into the following asm
```asm
G_M5589_IG03:
       488B5230             mov      rdx, qword ptr [rdx+48]   // irrelevant?
       488B12               mov      rdx, qword ptr [rdx]      // irrelevant?
       488B5208             mov      rdx, qword ptr [rdx+8]    // irrelevant?
       4983C010             add      r8, 16
       4963D1               movsxd   rdx, r9d                  // overwritten previous?
       498D14D0             lea      rdx, bword ptr [r8+8*rdx]
       488911               mov      bword ptr [rcx], rdx      // _pointer =  
       894108               mov      dword ptr [rcx+8], eax    // _length = length;
```
Seem to be 3 unnecessary memory loads?

/cc @jkotas

**Comment by mikedn** [comment]
Hmm, can't seem to reproduce this, I get code like:
```asm
G_M13287_IG02:
       4D85C9               test     r9, r9
       7439                 je       SHORT G_M13287_IG06

G_M13287_IG03:
       418B4108             mov      eax, dword ptr [r9+8]
       3BC2                 cmp      eax, edx
       723B                 jb       SHORT G_M13287_IG07
       2BC2                 sub      eax, edx
       413BC0               cmp      eax, r8d
       7234                 jb       SHORT G_M13287_IG07

G_M13287_IG04:
       498D4110             lea      rax, bword ptr [r9+16]
       4863CA               movsxd   rcx, edx
       488D0488             lea      rax, bword ptr [rax+4*rcx]
       488D4C2428           lea      rcx, bword ptr [rsp+28H]
       488901               mov      bword ptr [rcx], rax
       4489442430           mov      dword ptr [rsp+30H], r8d
```

What's before those useless moves you are seeing?

**Comment by benaadams** [comment]
Turned off force in-line and had the following code
```csharp
public ReadOnlySpan(T[] array, int start, int length)
{
    if (array == null || (array.Length - start < length) || (start | length) < 0)
        ThrowHelper.ThrowArgumentNullOrOutOfRangeException(array);

    _pointer = new ByReference<T>(ref Unsafe.Add(ref JitHelpers.GetArrayData(array), start));
    _length = length;
}
```
Full asm generated was
```asm
Marking ReadOnlySpan`1:.ctor(ref,int,int):this as NOINLINE because of unprofitable inline
Successfully inlined JitHelpers:GetArrayData(ref):byref (7 IL bytes) (depth 1) [below ALWAYS_INLINE size]
Successfully inlined Unsafe:Add(byref,int):byref (12 IL bytes) (depth 1) [aggressive inline attribute]
**************** Inline Tree
Inlines into 0600122E ReadOnlySpan`1:.ctor(ref,int,int):this
  [0 IL=0018 TR=000008 060002FB] [FAILED: does not return] ThrowHelper:ThrowArgumentNullOrOutOfRangeException(ref)
  [1 IL=0025 TR=000014 06003B07] [below ALWAYS_INLINE size] JitHelpers:GetArrayData(ref):byref
  [2 IL=0031 TR=000047 06003B0B] [aggressive inline attribute] Unsafe:Add(byref,int):byref
Budget: initialTime=222, finalTime=232, initialBudget=2220, currentBudget=2230
Budget: increased by 10 because of force inlines
Budget: initialSize=1362, finalSize=1362
; Assembly listing for method ReadOnlySpan`1:.ctor(ref,int,int):this
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T03] (  4,   4  )   byref  ->  rcx         this
;  V01 TypeCtx      [V01,T04] (  4,   4  )    long  ->  rdx        
;  V02 arg1         [V02,T00] (  7,   6  )     ref  ->   r8        
;  V03 arg2         [V03,T01] (  5,   5  )     int  ->   r9        
;  V04 arg3         [V04,T08] (  3,   3  )     int  ->  rax        
;  V05 loc0         [V05,T06] (  1,   2  )    long  ->  [rsp+0x28]  
;  V06 tmp1         [V06,T05] (  3,   6  )   byref  ->   r8        
;  V07 tmp2         [V07,T07] (  1,   2  )    long  ->  [rsp+0x20]  
;* V08 tmp3         [V08    ] (  0,   0  )  struct ( 8) zero-ref   
;  V09 tmp4         [V09,T09] (  2,   2  )   byref  ->  rdx         V08._value(offs=0x00) P-INDEP
;  V10 OutArgs      [V10    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;  V11 cse0         [V11,T02] (  5,   5  )    long  ->  rdx        
;
; Lcl frame size = 56

G_M5589_IG01:
       4883EC38             sub      rsp, 56
       4889542430           mov      qword ptr [rsp+30H], rdx
       8B442460             mov      eax, dword ptr [rsp+60H]

G_M5589_IG02:
       4D85C0               test     r8, r8
       0F8400000000         je       G_M5589_IG05
       458B5008             mov      r10d, dword ptr [r8+8]
       452BD1               sub      r10d, r9d
       443BD0               cmp      r10d, eax
       0F8C00000000         jl       G_M5589_IG05
       458BD1               mov      r10d, r9d
       440BD0               or       r10d, eax
       4585D2               test     r10d, r10d
       0F8C00000000         jl       G_M5589_IG05

G_M5589_IG03:
       488B5230             mov      rdx, qword ptr [rdx+48]   // irrelevant?
       488B12               mov      rdx, qword ptr [rdx]      // irrelevant?
       488B5208             mov      rdx, qword ptr [rdx+8]    // irrelevant?
       4983C010             add      r8, 16
       4963D1               movsxd   rdx, r9d                  // overwritte above?
       498D14D0             lea      rdx, bword ptr [r8+8*rdx]
       488911               mov      bword ptr [rcx], rdx      
       894108               mov      dword ptr [rcx+8], eax

G_M5589_IG04:
       4883C438             add      rsp, 56
       C3                   ret      

************** Beginning of cold code **************

G_M5589_IG05:
       498BC8               mov      rcx, r8
       E800000000           call     ThrowHelper:ThrowArgumentNullOrOutOfRangeException(ref)
       CC                   int3     

; Total bytes of code 96, prolog size 9 for method ReadOnlySpan`1:.ctor(ref,int,int):this
; ============================================================
```

**Comment by benaadams** [comment]
Will update to latest Jit to make sure is working as now

**Comment by mikedn** [comment]
I'll have to try again (it was inlined in my test) but it looks like another case of unused generic type context code being left around.

**Comment by mikedn** [comment]
Well, I get the following if the ctor is not inlined:
```asm
; Assembly listing for method System.ReadOnlySpan`1[Int32][System.Int32]:.ctor(ref,int,int):this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T04] (  4,   4  )   byref  ->  rcx         this
;  V01 arg1         [V01,T00] (  7,   7  )     ref  ->  rdx
;  V02 arg2         [V02,T01] (  7,   7  )     int  ->   r8
;  V03 arg3         [V03,T02] (  5,   5  )     int  ->   r9
;* V04 tmp0         [V04    ] (  0,   0  )  struct ( 8) zero-ref
;  V05 tmp1         [V05,T05] (  3,   6  )   byref  ->  rax
;  V06 tmp2         [V06,T06] (  2,   2  )   byref  ->  rax         V04._value(offs=0x00) P-INDEP
;  V07 OutArgs      [V07    ] (  1,   1  )  lclBlk (32) [rsp+0x00]
;  V08 cse0         [V08,T03] (  6,   6  )     int  ->  rax
;
; Lcl frame size = 40

G_M22710_IG01:
       4883EC28             sub      rsp, 40

G_M22710_IG02:
       4885D2               test     rdx, rdx
       7428                 je       SHORT G_M22710_IG06

G_M22710_IG03:
       8B4208               mov      eax, dword ptr [rdx+8]
       413BC0               cmp      eax, r8d
       722A                 jb       SHORT G_M22710_IG07
       412BC0               sub      eax, r8d
       413BC1               cmp      eax, r9d
       7222                 jb       SHORT G_M22710_IG07

G_M22710_IG04:
       488D4210             lea      rax, bword ptr [rdx+16]
       4963D0               movsxd   rdx, r8d
       488D0490             lea      rax, bword ptr [rax+4*rdx]
       488901               mov      bword ptr [rcx], rax
       44894908             mov      dword ptr [rcx+8], r9d

G_M22710_IG05:
       4883C428             add      rsp, 40
       C3                   ret

G_M22710_IG06:
       B903000000           mov      ecx, 3
       E886DDD6FF           call     System.ThrowHelper:ThrowArgumentNullException(int)

G_M22710_IG07:
       E8F9DCD6FF           call     System.ThrowHelper:ThrowArgumentOutOfRangeException()
       CC                   int3

; Total bytes of code 64, prolog size 4 for method System.ReadOnlySpan`1[Int32][System.Int32]:.ctor(ref,int,int):this
```

But which span are you talking about? The coreclr one or the corefx one?

The corefx one contains this code:
```C#
if (default(T) == null && array.GetType() != typeof(T[]))
```
I suspect that may generate the useless moves you are seeing. But then the corefx one doesn't use `ByReference` so I'm out of ideas...

**Comment by benaadams** [comment]
The IsReference is 
```csharp
public static bool IsReferenceOrContainsReferences<T>() => PerTypeValues<T>.IsReferenceOrContainsReferences;

public static bool IsReference<T>() => PerTypeValues<T>.IsReference;

public static bool IsReferenceCore(Type type)
{
    return type.GetTypeInfo().IsValueType;
}

public static class PerTypeValues<T>
{
    //
    // Latch to ensure that excruciatingly expensive validation check for constructing a Span around a raw pointer is done
    // only once per type.
    //
    public static readonly bool IsReferenceOrContainsReferences = IsReferenceOrContainsReferencesCore(typeof(T));

    public static readonly bool IsReference = IsReferenceCore(typeof(T));

    public static readonly T[] EmptyArray = new T[0];

    public static readonly IntPtr ArrayAdjustment = MeasureArrayAdjustment();

    // Array header sizes are a runtime implementation detail and aren't the same across all runtimes. (The CLR made a tweak after 4.5, and Mono has an extra Bounds pointer.)
    private static IntPtr MeasureArrayAdjustment()
    {
        T[] sampleArray = new T[1];
        return Unsafe.ByteOffset<T>(ref Unsafe.As<Pinnable<T>>(sampleArray).Data, ref sampleArray[0]);
    }
}
```
So does look like is generics 😢

**Comment by benaadams** [comment]
> But then the corefx one doesn't use ByReference so I'm out of ideas...

Was from PR https://github.com/dotnet/corefx/pull/16222

**Comment by benaadams** [comment]
Will see if `if (false && array.GetType() != typeof(T[]))` generates the same thing

**Comment by ahsonkhan** [comment]
@JosephTremoulet, any updates on this issue?

**Comment by JosephTremoulet** [comment]
I can see from the asm pastes above that @mikedn was looking at `ReadOnlySpan<int>`, but I can't see what the type argument was in @benaadams' dump.  I'm guessing it was a reference type, hence the generic sharing issues having a chance to surface in the one case but not the other.

And indeed when I look at ReadOnlySpan..ctor with current coreclr/jit sources, I see the extra indirections in `ReadOnlySpan<string>..ctor(ref string, int, int)` but not `ReadOnlySpan<int>..ctor(ref int, int, int)`.

As for where they come from:
  - ReadOnlySpan..ctor calls both `JitHelpers.GetArrayData` and `Unsafe.Add`
  - Both of these are shared-to-shared calls, so each callsite gets a generics context lookup
  - The context lookup involves the same lazy initialization pattern seen in #10051 (`tok = *(*callerCtx)); if (tok == null) { tok = GetContextToken(...); }`)
  - Neither `JitHelpers.GetArrayData` nor `Unsafe.Add` actually use their context parameter
  - The inliner notices this and avoids setting `lvaGenericsContextUsed` (thanks to #9756), but doesn't rip out the context lookup code (and doesn't have a pointer to that code handy)
  - `fgLocalVarLiveness` recognizes the helper calls in the `qmark`s as useless and removes them, but only after morph has expanded the `qmark`s to explicit flow
  - we wind up with a conditional branch around an empty block, which persists all the way until `fgUpdateFlowGraph`
  - once the blocks are collapsed and the conditional jump is removed, the expression feeding the conditional jump could be removed, but nothing in that part of the phase list is doing that sort of analysis (the compare against zero gets removed, but not the load feeding the compare)

So the loads are there because they used to feed compares that were used to determine whether to make helper calls which have been removed.  Each of the 2 occurrences has two loads, but the first load in the one is redundant with the first load in the other, so we end up with 3 loads in the emitted asm.

This might be more motivation to do some sort of deferred lowering for this context lookup control flow (similar to #10051), or perhaps `fgUpdateFlowGraph` or whatever removes the compare could be extended to remove the load as well, since it's marked `GTF_IND_NONFAULTING`...

**Comment by JosephTremoulet** [comment]
I was wrong when I said "nothing in that part of the phase list is doing that sort of analysis" -- the subsequent `fgLocalVarLiveness` pass does identify and remove dead local stores, but it's getting blocked here because the final load has `GTF_ASG` set for some reason...

**Comment by JosephTremoulet** [comment]
The `GTF_ASG` gets added during CSE; we have this coming in, and the indir in question is the 3rd-level one, [000029]:

```
***** BB04, stmt 4
     ( 10,  9)              [000032] ------------             *  stmtExpr  void  (IL 0x01C...0x02E)
N008 ( 10,  9)              [000029] ------------             |  /--*  indir     long   <l:$340, c:$380>
N006 (  1,  1)              [000027] ------------             |  |  |  /--*  const     long   8 $282
N007 (  8,  7)              [000028] -------N----             |  |  \--*  +         long   $2c1
N005 (  7,  6) CSE #03 (def)[000026] #-----------             |  |     \--*  indir     long   $301
N004 (  4,  4) CSE #02 (def)[000025] #-----------             |  |        \--*  indir     long   $300
N002 (  1,  1)              [000023] ------------             |  |           |  /--*  const     long   48 $280
N003 (  2,  2)              [000024] -------N----             |  |           \--*  +         long   $2c0
N001 (  1,  1)              [000021] ------------             |  |              \--*  lclVar    long   V01 TypeCtx      u:2 $c0
N010 ( 10,  9)              [000031] -A------R---             \--*  =         long   <l:$340, c:$380>
N009 (  1,  1)              [000030] D------N----                \--*  lclVar    long   V05 loc0         d:2 <l:$340, c:$380>
```

The inner expression marked CSE #<cse>03 becomes a comma that includes an assign to the temp introduced for CSEing it, and the load load inherits the `GTF_ASG` flag:

```
N012 ( 11, 10)              [000029] -A----------             /--*  indir     long   <l:$340, c:$380>
N010 (  1,  1)              [000027] ------------             |  |  /--*  const     long   8 $282
N011 (  9,  8)              [000028] -A-----N----             |  \--*  +         long   $2c1
N008 (  1,  1)              [000179] ------------             |     |  /--*  lclVar    long   V11 cse0          $301
N009 (  8,  7)              [000180] -A----------             |     \--*  comma     long   $301
N005 (  7,  6)              [000026] #-----------             |        |  /--*  indir     long   $301
N004 (  4,  4) CSE #02 (def)[000025] #-----------             |        |  |  \--*  indir     long   $300
N002 (  1,  1)              [000023] ------------             |        |  |     |  /--*  const     long   48 $280
N003 (  2,  2)              [000024] -------N----             |        |  |     \--*  +         long   $2c0
N001 (  1,  1)              [000021] ------------             |        |  |        \--*  lclVar    long   V01 TypeCtx      u:2 $c0
N007 (  7,  6)              [000178] -A------R---             |        \--*  =         long   $VN.Void
N006 (  1,  1)              [000177] D------N----             |           \--*  lclVar    long   V11 cse0          $301
N014 ( 11, 10)              [000031] -A------R---             *  =         long   <l:$340, c:$380>
N013 (  1,  1)              [000030] D------N----             \--*  lclVar    long   V05 loc0         d:2 <l:$340, c:$380>
```

The tree persists in that form until rationalize, where we unpack the comma but don't clean up the `GTF_ASG` that no longer applies to the outer indir:

```
------------ BB04 [01C..02E) -> BB06 (cond), preds={BB03} succs={BB05,BB06}
     ( 11, 10) [000032] ------------                il_offset void   IL offset: 28
N001 (  1,  1) [000021] ------------       t21 =    lclVar    long   V01 TypeCtx      u:2 $c0
N002 (  1,  1) [000023] ------------       t23 =    const     long   48 $280
                                                 /--*  t21    long   
                                                 +--*  t23    long   
N003 (  2,  2) [000024] -------N----       t24 = *  +         long   $2c0
                                                 /--*  t24    long   
N004 (  4,  4) [000025] #-----------       t25 = *  indir     long   $300
                                                 /--*  t25    long   
N005 (  7,  6) [000026] #-----------       t26 = *  indir     long   $301
                                                 /--*  t26    long   
N007 (  7,  6) [000178] DA----------             *  st.lclVar long   V11 cse0         
N008 (  1,  1) [000179] ------------      t179 =    lclVar    long   V11 cse0          $301
N010 (  1,  1) [000027] ------------       t27 =    const     long   8 $282
                                                 /--*  t179   long   
                                                 +--*  t27    long   
N011 (  9,  8) [000028] -A-----N----       t28 = *  +         long   $2c1
                                                 /--*  t28    long   
N012 ( 11, 10) [000029] -A----------       t29 = *  indir     long   <l:$340, c:$380>
                                                 /--*  t29    long   
N014 ( 11, 10) [000031] DA----------             *  st.lclVar long   V05 loc0         d:2
```

@dotnet/jit-contrib, is rationalize just oblivious to the flags, or do we have some flag cleanup that's just missing this case?

**Comment by JosephTremoulet** [comment]
Rationalize is not updating the flags, but updating it to patch up `GTF_ASG` as it goes cleans this up and from looking at asm diffs removes a smattering of other dead sequences as well.  I'll polish up those changes and send out a PR.
